### ISSUE TITLE
iOS PR proposal: restricting pan movement

### ISSUE THREAD

**Issue by mflint** [maplibre/maplibre-native#2115]
**Is your feature request related to a problem? Please describe.**
I'd like to restrict the distance that the customer can pan from a particular point.

From reading questions on StackOverflow and other places, the standard response seems to be:

> implement the shouldChange delegate, set a different mapView.centerCoordinate and return false

This results in quite a jumpy experience, and panning along the boundary can feel _sticky_. It also feels wrong to be abusing the delegate callback like that, by mutating map state while handling a delegate callback.

**Describe the solution you'd like**
I've made a fork which:
* adds `- (CLLocationCoordinate2D)mapView:(MLNMapView *)mapView oldCoordinate:(CLLocationCoordinate2D)oldCoordinate limitPanCoordinate:(CLLocationCoordinate2D)panCoordinate` to `MLNMapViewDelegate.h`
  * this supplies the previous centre map coordinate and the proposed new centre coordinate (that the user is trying to pan to) and allows the delegate to offer an alternate pan coordinate instead
* two changes to `-[MLNMapView handlePanGesture:]` to call that new delegate function, if implemented

As a trivial example, this only allows the user to pan the map north/south, by taking the latitude value from the new coordinate and longitude value from the original coordinate:

```swift
func mapView(_ mapView: MLNMapView,
             oldCoordinate: CLLocationCoordinate2D,
             limitPanCoordinate panCoordinate: CLLocationCoordinate2D) -> CLLocationCoordinate2D {
	// only allow north/south panning
	return CLLocationCoordinate2D(latitude: panCoordinate.latitude, longitude: oldCoordinate.longitude)
}
```

In my app, I wanted to restrict panning within 2km of a location - so the limit of the pan is a circle. If I try to pan further, then my delegate finds the closest point on the circle, and returns that coordinate instead.

**Describe alternatives you've considered**
Setting a new location from inside the `shouldChangeFrom:to:` delegate function feels like it's abusing the contract of what a delegate should do.

**Additional context**
I'm happy to continue maintaining my fork, but I would like to offer the idea upstream.

My local change has no automated tests, but it could be tidied-up into a PR.

**Comment by louwers** [comment]
> panning along the boundary can feel sticky

Yes maybe this should be made a bit easier.

I don't want you to have to maintain a fork to get the features you want. If there are more people that would benefit from this functionality I would say we should go ahead and implement it.

Although the MLNMapView and delegate are quite unwieldy already...

**Comment by mflint** [comment]
> Although the MLNMapView and delegate are quite unwieldy already...

You are correct, the delegate is large... but, helpfully, it's already annotated with `// MARK:` sections, so could be split into separate protocols.
